crates/matrix-sdk/src/room/power_levels.rs

```diff
@@ -0,0 +1,289 @@
+            users_default: Some(value.users_default.into()),
+            room_name: value
+                .events
+                .get(&StateEventType::RoomName.into())
+                .map(|v| (*v).into())
+                .or(Some(value.state_default.into())),
+            room_avatar: value
+                .events
+                .get(&StateEventType::RoomAvatar.into())
+                .map(|v| (*v).into())
+                .or(Some(value.state_default.into())),
+            room_topic: value
+                .events
+                .get(&StateEventType::RoomTopic.into())
+                .map(|v| (*v).into())
+                .or(Some(value.state_default.into())),
+        }
+    }
+}
+
+pub(crate) trait RoomPowerLevelsExt {
+    /// Applies the updated settings to the power levels. Any levels that are
+    /// `None` will remain unchanged.
+    fn apply(&mut self, settings: RoomPowerLevelSettings) -> anyhow::Result<(), anyhow::Error>;
+    /// Updates the power level for a specific state event. If the new level is
+    /// the default level, then the event will be removed.
+    fn update_state_event(
+        &mut self,
+        event_type: StateEventType,
+        new_level: i64,
+    ) -> anyhow::Result<(), anyhow::Error>;
+}
+
+impl RoomPowerLevelsExt for RoomPowerLevels {
+    fn apply(&mut self, settings: RoomPowerLevelSettings) -> anyhow::Result<(), anyhow::Error> {
+        if let Some(ban) = settings.ban {
+            self.ban = ban.try_into()?;
+        }
+        if let Some(invite) = settings.invite {
+            self.invite = invite.try_into()?;
+        }
+        if let Some(kick) = settings.kick {
+            self.kick = kick.try_into()?;
+        }
+        if let Some(redact) = settings.redact {
+            self.redact = redact.try_into()?;
+        }
+        if let Some(events_default) = settings.events_default {
+            self.events_default = events_default.try_into()?;
+        }
+        if let Some(state_default) = settings.state_default {
+            self.state_default = state_default.try_into()?;
+        }
+        if let Some(users_default) = settings.users_default {
+            self.users_default = users_default.try_into()?;
+        }
+        if let Some(room_name) = settings.room_name {
+            self.update_state_event(StateEventType::RoomName, room_name)?;
+        }
+        if let Some(room_avatar) = settings.room_avatar {
+            self.update_state_event(StateEventType::RoomAvatar, room_avatar)?;
+        }
+        if let Some(room_topic) = settings.room_topic {
+            self.update_state_event(StateEventType::RoomTopic, room_topic)?;
+        }
+
+        Ok(())
+    }
+
+    fn update_state_event(
+        &mut self,
+        event_type: StateEventType,
+        new_level: i64,
+    ) -> anyhow::Result<(), anyhow::Error> {
+        let new_level = new_level.try_into()?;
+        if new_level == self.state_default {
+            self.events.remove(&event_type.into());
```

Comment:
I'm wondering if this is a sensible option here as I have noticed that we create rooms with a default value for e.g. the avatar that also matches the `state_default`.